validator/client/src/main/java/tech/pegasys/teku/validator/client/doppelganger/DoppelgangerDetector.java

```diff
@@ -0,0 +1,297 @@
+    this.spec = spec;
+    this.timeProvider = timeProvider;
+    this.genesisDataProvider = genesisDataProvider;
+    this.checkDelay = checkDelay;
+    this.timeout = timeout;
+    this.maxEpochs = maxEpochs;
+  }
+
+  public SafeFuture<Map<UInt64, BLSPublicKey>> performDoppelgangerDetection(
+      final Set<BLSPublicKey> pubKeys) {
+    if (pubKeys.isEmpty()) {
+      LOG.info("Skipping doppelganger detection: No public keys to check");
+      return SafeFuture.completedFuture(new HashMap<>());
+    }
+    DoppelgangerDetectionTask doppelgangerDetectionTask =
+        new DoppelgangerDetectionTask(timeProvider.getTimeInSeconds(), pubKeys);
+    return doppelgangerDetectionTask.performDoppelgangerDetectionTask();
+  }
+
+  private class DoppelgangerDetectionTask {
+
+    private volatile Cancellable doppelgangerDetectionTask;
+    private Optional<SafeFuture<Map<UInt64, BLSPublicKey>>> doppelgangerCheckFinished =
+        Optional.empty();
+    private final UInt64 startTime;
+    private final Set<BLSPublicKey> pubKeys;
+    private Optional<UInt64> epochAtStart = Optional.empty();
+
+    public DoppelgangerDetectionTask(final UInt64 startTime, final Set<BLSPublicKey> pubKeys) {
+      this.startTime = startTime;
+      this.pubKeys = pubKeys;
+    }
+
+    protected synchronized SafeFuture<Map<UInt64, BLSPublicKey>>
+        performDoppelgangerDetectionTask() {
+      statusLog.doppelgangerDetectionStart(
+          pubKeys.stream().map(BLSPublicKey::toAbbreviatedString).collect(Collectors.toSet()));
+      doppelgangerCheckFinished = Optional.of(new SafeFuture<>());
+      epochAtStart = Optional.empty();
+      doppelgangerDetectionTask =
+          asyncRunner.runWithFixedDelay(
+              () -> performDoppelgangerCheck(pubKeys),
+              checkDelay,
+              checkDelay,
+              throwable ->
+                  LOG.error(
+                      "Error while checking validators doppelganger. The check could not be performed correctly.",
+                      throwable));
+      return doppelgangerCheckFinished.get();
+    }
+
+    private synchronized SafeFuture<Void> stopDoppelgangerDetector(
+        final Map<UInt64, BLSPublicKey> detectedDoppelgangers) {
+      if (doppelgangerCheckFinished.isEmpty()) {
+        throw new RuntimeException("Doppelganger Detection is already stopped");
+      }
+      epochAtStart = Optional.empty();
+      doppelgangerCheckFinished.get().complete(detectedDoppelgangers);
+      doppelgangerCheckFinished = Optional.empty();
+      return SafeFuture.fromRunnable(
+          () -> Optional.ofNullable(doppelgangerDetectionTask).ifPresent(Cancellable::cancel));
+    }
+
+    private SafeFuture<Void> performDoppelgangerCheck(final Set<BLSPublicKey> pubKeys) {
+      if (timeProvider
+          .getTimeInSeconds()
+          .minus(startTime)
+          .isGreaterThanOrEqualTo(timeout.toSeconds())) {
+        LOG.warn("Doppelganger Detection timeout");
```

Comment:
When i see this as a user, what do I do? As a warning, it's going to be pretty visible, and users will want to know what actions they can take...

Reply:
Done. Added more details and moved the log to the `StatusLogger`